### PR TITLE
refactor: Add `EtherAmount` to simplify operating with ETH and wstETH [WIP]

### PR DESCRIPTION
**Pull request by MeLlamoPablo**
This PR introduces an abstract class `BigAmount` and two children: `EtherAmount` and `Erc20Amount`. They are intended to become the primary way we operate with amount values to improve code consistency, safety and maintainability.

## Definition

An amount is a numeric value representing the quantity of a given token.

* `2 ETH` is an amount, represented by `EtherAmount`
* `0.5 zkSync Long Token` is an amount, represented by `Erc20Amount`
* `2% deposit fee` is NOT an amount because it does not refer to a token quantity.
* `1.1 ETH per wstETH` is NOT an amount either.

## Motivation and design goals

* Unify operating with ETH and wstETH.

  We operate with `wstETH` conversion frequently within our codebase, resulting in varaibles named `amountInEth` and `amountInWstEth`, increasing code verbosity and having to handle many loading states. We convert back and forth too many times.

  `EtherAmount` is designed to abstract this and work with both units. You can instantiate it with `.fromEth` and `.fromWstEth`. You can operate with them seamlessly, even if they are different in denomination:

  ```ts
  const a = EtherAmount.fromEth(depositAmount, this.root)
  const b = EtherAmount.fromWstEth(depositFee, this.root)

  const balanceIncrease = depositAmount.sub(depositFee)
  ```

* Unify operating with `BigNumber` and `string`

  Across the codebase we use `BigNumber` and the "formatted" `string` representation. Operating with `BigNumber` is precise, but the formatted string represents "token units" which is more convenient.

  This leads to confusion, and variables named `amountBN` and `amount/amountString/amountFormat`.

  A design goal is to combine both of these properties. `BigAmount`s have information about the decimals of the token they represent, and so we can operate with them using `BigNumber` precision but still be able to format as a string whenever we need to show the value in the UI.

  We should only need to convert a `BigAmount` to a string when showing information in the UI.

* Improve type safety.

  Since `BigAmount`s refer to specific tokens, we can increase the type safety of our operations:

  * Adding `EtherAmount + EtherAmount` makes sense and is allowed
  * Adding `EtherAmount + Erc20Amount` doesn't makes sense and is not allowed
  * Multiplying `EtherAmount * BigNumber` makes sense and is allowed
  * Multiplying `EtherAmount * EtherAmount` doesn't make sense and isn't allowed

* Abstract away `BigNumber`

  The goal is to completely wrap `BigNumber` so that we don't have to interact with it directly. Theoretically, we could even switch the underlying `BigNumber` library if we wanted and it would still work.

  This allows us to easily extend its functionality. For example, instead of throwing errors on division-by-zero, we safely default to 0 and log an error to the console.

  The only two cases where we should need to unwrap to `BigNumber` are:

  * Legacy code compatibility (we can't migrate all the codebase at once)
  * Sending `BigNumber` to the blockchain
    * In the future, we could add a ESLint rule to ensure that `inEthBN`/`inWstEthBN` only get called inside `sendTransaction`

* Batteries-included operations

  Currently, whenever we want to operate with amounts, we sometimes need to get additional information from the stores: call `BalancerStore` for conversion and call the token store for the amount of decimals.

  `EhterAmount`/`Erc20Amount` are designed to self-contain all of their dependencies. After you instantiate them, you don't need to use any external store to operate or display data:

  ```ts
  // In the store
  const amount = EtherAmount.fromWstEth(BigNumber.from(100_000), this.root)
  const half = amount.div(BigNumber.from(2))
  
  // In the UI
  half.toFormattedEthString() // Converts wstETH -> ETH and formats at the same time.
  ```

## Scope of this PR

This PR:

* Introduces the `BigAmount`, `EtherAmount` and `Erc20Amount` classes
  * The `Erc20Amount` doesn't have usages
* Refactors `DepositRecordStore` and `WithdrawHook` store to return `EtherAmount` directly

  This showcases how it should be used everywhere: we convert `BigNumber` to `EtherAmount` directly at the source, as soon as the blockchain returns the data to us.
* Refactors `DepositStore` to showcase how it would be use within a store
* Other changes to other files to ensure it doesn't break.

### CODE REVIEW

**Comment by MeLlamoPablo** [comment]
@chrisling-dev I think this is ready to receive feedback, even though the tests are failing. Before putting more work into this I'd like to make sure that we're on the same page.

**Comment by chrisling-dev** [comment]
Hey @MeLlamoPablo could you help rebase this PR for review

**Comment by MeLlamoPablo** [comment]
Done @chrisling-dev

**Review by chrisling-dev — CHANGES_REQUESTED** [review]
This PR is too large for me to review confidently, could you break it down into smaller PRs? I would at least have:

- Deposit: 1 PR
- Withdraw: 1 PR
- Trade: 1 PR

The idea is that each PR should not affect more than 1 feature so it's easier to test, because this PR is quite complex

Also try to avoid any irrelevant changes like the comment in Layout.tsx

File: apps/frontend/core-dapp/src/utils/fraction-utils.ts [hunk 1]
```diff
@@ -33,7 +34,10 @@ export function fraction(num: BigNumber, den: BigNumber): Fraction {
   return [num.div(commonDivisor), den.div(commonDivisor)] as Fraction
 }
 
-export function mul(number: BigNumber, [num, den]: Fraction): BigNumber {
+export function mul(number: BigNumber, f: Fraction): BigNumber
+export function mul<T extends BigAmount>(number: T, f: Fraction): T
+
+export function mul(number: BigNumber | BigAmount, [num, den]: Fraction): BigNumber | BigAmount {
```
[the remark this answers is not in the document]
Comment (MeLlamoPablo) [reply apps/frontend/core-dapp/src/utils/fraction-utils.ts:40]: These are [Typescript Function Overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads). They provide signatures more precise than the implementation signature.

In this case, the implementation signature can take either `BigNumber` or `BigAmount` and can return either `BigNumber` or `BigAmount`. If we had only that, we'd have to cast the result to what we need every time.

With the two above overloads we can specify more precise types:

* If the input is `BigNumber`, the result is `BigNumber`
* If the input is a child of `BigAmount`, the result is that child of `BigAmount`

**Comment by MeLlamoPablo** [comment]
@chrisling-dev opened #1310 with just the `EtherAmount` class and its tests. This PR is intended to hold all WIP changes but not intended to be merged. After that one is merged I'll continue splitting this one.

**Comment by MeLlamoPablo** [comment]
> Also try to avoid any irrelevant changes like the comment in Layout.tsx

This is not an irrelevant change, the build would fail without it due to the ESLint change

**Pull request closed without merging**